aten/src/ATen/native/cpu/DistanceOpsKernel.cpp

```diff
@@ -104,7 +104,11 @@ struct Dist {
 
   // Special general pnorm derivative if p is less than two
   struct lttdist_calc {
-    static inline Vec backward(const Vec& diff, const scalar_t grad, const scalar_t dist, const Vec& p) { return dist == 0.0 ? Vec(0) : sign(diff) * diff.abs().pow(p - Vec(1)) * Vec(grad) / Vec(dist).pow(p - Vec(1)); }
+    static inline Vec backward(const Vec& diff, const scalar_t grad, const scalar_t dist, const Vec& p) {
+      Vec result = (dist == 0.0) ? Vec(0) : (sign(diff) * diff.abs().pow(p - Vec(1)) * Vec(grad) / Vec(dist).pow(p - Vec(1)));
+      result = Vec::blendv(result, Vec(0), (diff == Vec(0)) & (p < Vec(1)));
+      return result;
+    }
```

Comment:
The gradients in here is defined for the
`cdist(x, y) = \sum_i (|x_i - y_i|^p)^{1/p}`. For values 0 < p < 1, this function is not a norm, and the underlying space is not a metric space. And it is not differentiable at zero, which means that the value of `result` could be made arbitrary large.
However, if for values 0 < p < 1 we ignore the p-root and consider:
`cdist(x, y) = \sum_i |x_i - y_i|^p`, now this function is a metric, so it does indeed measure distance.

Reply:
Another argument, the derivative limits from the left and right have huge values and of opposite sign. It means, that if we are to run gradient descent, it is very likely to zig-zag around zero without convergence.

Reply:
The logic that we use for these is the following (similar to what we do for norm): The function is convex and admits a subgradient at 0. And we use the min-norm subgradient when multiple ones exists to ensure that it provides a descend direction. In this case 0.

Reply:
Case `p=0` is special, if we adopt the metric
`cdist(x, y)_p = \sum_i |x_i - y_i|^p`, then
`cdist(x, y)_0 = \sum_i {1 if |x_i - y_i| > 0}`

Reply:
Ah, OK, there is yet another definition of 0-"norm". More on the general case [here](https://en.wikipedia.org/wiki/Lp_space#When_0_%3C_p_%3C_1).
@albanD , your thoughts?

Reply:
For `p=0` we should return 0 gradient all the time IIRC (piecewise constant function).
This is not properly special cased for cdist? There is a zdist above in this file but maybe this is not doing the right thing.

Reply:
No, the function is actually concave for `0<p<1` :(

Reply:
The problem is not with returning zero, but rather in what neighborhood to return it. So, the convention to return zero once in `eps` neighborhood limited by `1e-45`?

Reply:
It's not always indeed but I would argue that this is fine as:
- This is true for the case where these is a single element as the abs() function is.
- When there are more elements, this remains below the whole function and touches it at the point of interest.
- We don't want to provide a non-zero gradient at zero for these functions as we would have to pick the sign arbitrarily. And we don't want to return nan as it is not user-friendly.

Can you think of another definition we can use here?

Reply:
> The problem is not with returning zero, but rather in what neighborhood to return it. So, the convention to return zero once in eps neighborhood limited by 1e-45?

Ho, we don't have an answer to that I'm afraid.
Currently for all the norm-like functions, only `0` will have a custom value but values very close to 0 can generate inf/nan due to numerical precision issues (see the issue I liked above).

`1e-45` doesn't sound good (at least for float32) as you will get inf/nan way before reaching that.
The two options I can see are:
- Arbitrarily pick a range of value where we override the gradient (note that this range will be dtype dependant) but that might not solve all the problems and can be flaky.
- Inspect the computed gradient while knowing the formula and change inf/nan generated by the global formula if we detect that this is due to numerical precision (because the final value is close to 0)? But that would require some work to make this work properly for complex formulas like "norm" that can create inf/nan for many reasons beyond just the value being close to 0.

Reply:
Nothing against zero at zero, but the chance of this exact zero has probability zero. The function is not differentiable at zero, which means it is likely to get numerical issues at values close to zero. One possible resolution of this is by bounding the gradients from above by some max value, but then this bound depends on the inputs' dimension. Do you think this is feasible?

Reply:
Sorry, Alban, there is a delay with which our replies appear. Then what do you think about making a warning in the documentation of these norm-like functions with these specific values? Something like: "expect numerical issues around zero, consider bounding the gradients as a possible workaround", or something along these line?

Reply:
> this exact zero has probability zero. 

Not in the amazing neural network world where half of your network contains `relu` that set things to 0 all the time haha
Also for distance functions, I wouldn't say that people do `dist(a, a)` with probability 0 (maybe not on purpose, but I have definitely already seen this).

> One possible resolution of this is by bounding the gradients from above by some max value

For p<1, this might be ok but mathematically, the gradient does go to infinity when you're close to 0 so I don't see why you shouldn't get such very large value? The analogy here for me is sqrt that will return larger and larger values as you get closer to 0 and it will eventually give you inf.
For p>1, this would actually be wrong as the gradient is, in theory, smoothly converging towards 0 at 0 and only numerical inaccuracies make it grow.

Reply:
Ok, I am convinced :) Let the user decide then, he is smart enough to understand the issues. Sorry for the confusion!

Reply:
:)
There is still an issue for p>1 for values very close to 0 though. But this is beyond the scope of this PR. But if you have ideas on how we can approach that, I am very interested!

Reply:
Yeah, it seem we can rearrange things a bit to reduce the effect, but it is going to be bc-breaking I presume.